src/os_db.ml

```diff
@@ -1531,70 +1531,136 @@ module User =
+                                      (List.map
+                                         (function
+                                          | `Text text -> text
+                                          | `Var (_varname, false, _) ->
+                                              let () = incr i in
+                                              let () = incr j in
+                                              "$" ^
+                                                (string_of_int j.contents)
+                                          | `Var (_varname, true, _) ->
+                                              let param =
+                                                List.nth params i.contents in
+                                              let () = incr i in
+                                              "(" ^
+                                                ((String.concat ","
+                                                    (List.map
+                                                       (fun _ ->
+                                                          let () = incr j in
+                                                          "$" ^
+                                                            (string_of_int
+                                                               j.contents))
+                                                       param))
+                                                   ^ ")")) split) in
+                                  let params = List.flatten params in
+                                  let name =
+                                    "ppx_pgsql." ^
+                                      (Digest.to_hex (Digest.string query)) in
+                                  let hash =
+                                    try PGOCaml.private_data dbh
+                                    with
+                                    | Not_found ->
+                                        let hash = Hashtbl.create 17 in
+                                        (PGOCaml.set_private_data dbh hash;
+                                         hash) in
+                                  let is_prepared = Hashtbl.mem hash name in
+                                  PGOCaml.bind
+                                    (if not is_prepared
+                                     then
+                                       PGOCaml.bind
+                                         (PGOCaml.prepare dbh ~name ~query ())
+                                         (fun () ->
+                                            Hashtbl.add hash name true;
+                                            PGOCaml.return ())
+                                     else PGOCaml.return ())
+                                    (fun () ->
+                                       PGOCaml.execute_rev dbh ~name ~params
+                                         ()))
+                                 (fun _rows -> PGOCaml.return ()) in
```

Comment:
Is it particularly difficult to share code between the `then` and the `else` branch? I'm having trouble finding the differences and the commonalities between them.

Reply:
This code is generated by the PGOCaml ppx from `os_db.ppx.ml`